### PR TITLE
fix (core/pdump): Fix Pdump write and load thru TC Cherry picks

### PR DESCRIPTION
**Pull request by acidmanifesto**
TC Cherry picks were over 30+. I am not listing it all but i am listing authors of said relations that this pr is based on.

Playerdump is now aligned about 99% with trintiycore's current which is functional where ours is not.

Co-Authored-By: Brian <runningnak3d@gmail.com>
Co-Authored-By: joschiwald <736792+joschiwald@users.noreply.github.com>
Co-Authored-By: Shauren <shauren.trinity@gmail.com>
Co-Authored-By: SnapperRy <19622383+SnapperRy@users.noreply.github.com>
Co-Authored-By: Ariel Silva <ariel-@users.noreply.github.com>
Co-Authored-By: Nick <31777+paradox@users.noreply.github.com>
Co-Authored-By: r00ty-tc <r00ty-tc@users.noreply.github.com>
Co-Authored-By: Jorge García <259712+xurxogr@users.noreply.github.com>
Co-Authored-By: Kaelima <kaelima@live.se>
Co-Authored-By: Shocker <43253032+shockerqt@users.noreply.github.com>
Co-Authored-By: Giacomo Pozzoni <giacomopoz@gmail.com>

## Changes Proposed:
-  PDump write and PDump load now works

## Issues Addressed:

- Closes https://github.com/azerothcore/azerothcore-wotlk/issues/7346

## SOURCE:

-- Trinitycore 335

## Tests Performed:

- builds
- performs
## How to Test the Changes:

1. backed up my warrior tach with .pdump write warrior tach ( .pdump write FILENAME CHARACTER NAME)
image

2. Deleted character from character screen that i backed up with pdump write. 
image
image

3. Load up the character with .pdump load warrior test (.pdump load FILENAME ACCOUNT NAME)
image

4. Exit to player select screen
image

5. select character which prompt for rename.
image

6. observer character had original gear and location and stats
image

## Known Issues and TODO List:

- [ ]
- [ ]

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

File: src/common/Common.h [hunk 1]
```diff
@@ -51,6 +51,8 @@
 #define STRINGIZE(a) #a
 
 #define MAX_NETCLIENT_PACKET_SIZE (32767 - 1)               // Client hardcap: int16 with trailing zero space otherwise crash on memory free
+inline unsigned long atoul(char const* str) { return strtoul(str, nullptr, 10); }
+inline unsigned long long atoull(char const* str) { return strtoull(str, nullptr, 10); }
```
Comment (Kitzunu) [inline src/common/Common.h:55]: no. use StringTo

File: src/server/game/Tools/PlayerDump.cpp [hunk 2]
```diff
@@ -20,378 +20,713 @@
...
+    GUID_TYPE_PET,
+    GUID_TYPE_MAIL,
+    GUID_TYPE_ITEM,
+
+    // 64 bit long guids
+    GUID_TYPE_EQUIPMENT_SET,
+
+    // special types
+    GUID_TYPE_NULL // set to null
+};
```
Comment (Kitzunu) [inline src/server/game/Tools/PlayerDump.cpp:45]:
```suggestion
enum class GuidType : uint8
{
    // 32 bit long guids
    Account,
    Char,
    Pet,
    Mail,
    Item,

    // 64 bit long guids
    EquipmentSet,

    // special types
    Null // set to null
};
```

File: src/server/game/Tools/PlayerDump.cpp [hunk 3]
```diff
@@ -20,378 +20,713 @@
...
+
+// for RAII
+struct FileCloser
+{
+    void operator()(FILE* f) const
+    {
+        if (f)
+            fclose(f);
```
Comment (Kitzunu) [inline src/server/game/Tools/PlayerDump.cpp:53]: brackets and weird indent

**Review by Kitzunu — CHANGES_REQUESTED** [review]

File: src/server/game/Tools/PlayerDump.cpp (hunk 2 above)
Comment (acidmanifesto) [reply src/server/game/Tools/PlayerDump.cpp:45]: if i rename it to what you suggest i will need to rename it in a few other spots as well.  Would it be necessary for me to rename them?

Comment (Kitzunu) [reply src/server/game/Tools/PlayerDump.cpp:45]: I mean no, but enum class is preferred https://www.azerothcore.org/wiki/cpp-code-standards#enum-vs-enum-class

What you would need to rename them to in the rest of the code would be GuidType::Account for example

Comment (acidmanifesto) [reply src/server/game/Tools/PlayerDump.cpp:45]: i will do the changing up after the compile testings are done incase i need to address anything else. Thanks @Kitzunu !

**Comment by benjymansy123** [comment]
Tested, works as expected.

**Review by Kitzunu — APPROVED** [review]

**Comment by acidmanifesto** [comment]
Retested with Latest changes and does still work as intended. Thank you @Kitzunu for your assistance

**Pull request merged**